quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java

```diff
@@ -558,124 +551,62 @@ private static void outputDisabledProperties(Set<String> properties, boolean bui
+                return;
             }
-        }
-
-        //check for defined quarkus raw build properties for UserStorageProvider extensions
-        if (QuarkusPropertiesConfigSource.getConfigurationFile() != null) {
-            Optional<ConfigSource> quarkusPropertiesConfigSource = getConfig().getConfigSource(QuarkusPropertiesConfigSource.NAME);
-
-            if (quarkusPropertiesConfigSource.isPresent()) {
-                Map<String, String> foundQuarkusBuildProperties = findSupportedRawQuarkusBuildProperties(quarkusPropertiesConfigSource.get().getProperties().entrySet());
-
-                //only check if buildProps are found in quarkus properties file.
-                if (!foundQuarkusBuildProperties.isEmpty()) {
-                    Optional<ConfigSource> persistedConfigSource = getConfig().getConfigSource(PersistedConfigSource.NAME);
-
-                    if(persistedConfigSource.isPresent()) {
-                        for(String key : foundQuarkusBuildProperties.keySet()) {
-                            if (notContainsKey(persistedConfigSource.get(), key)) {
-                                //if persisted cs does not contain raw quarkus key from quarkus.properties, assume build is needed as the key is new.
-                                return true;
-                            }
-                        }
-
-                        //if it contains the key, check if the value actually changed from the persisted one.
-                        return hasAtLeastOneChangedBuildProperty(foundQuarkusBuildProperties, persistedConfigSource.get().getProperties().entrySet());
-                    }
-                }
+            // since we're presisting all quarkus values, this may leak some runtime information - we don't want
+            // to capture expanded expressions that may be referencing environment variables
+            String stringValue = value.getValue();
+            if (quarkus && value.getRawValue() != null) {
+                stringValue = value.getRawValue();
             }
-        }
+            properties.put(name, stringValue);
+        });
 
-        return false;
-    }
+        // the following should be ignored when output the optimized check message
+        // they are either not set by the user, or not properly initialized
 
-    private static boolean hasAtLeastOneChangedBuildProperty(Map<String, String> foundQuarkusBuildProperties, Set<Map.Entry<String, String>> persistedEntries) {
-        for(Map.Entry<String, String> persistedEntry : persistedEntries) {
-            if (foundQuarkusBuildProperties.containsKey(persistedEntry.getKey())) {
-                return isChangedValue(foundQuarkusBuildProperties, persistedEntry);
-            }
+        for (File jar : getProviderFiles().values()) {
+            properties.put(String.format(KC_PROVIDER_FILE_PREFIX + "%s.last-modified", jar.getName()), String.valueOf(jar.lastModified()));
```

Comment:
How do we check change for custom providers that need a rebuild? From reading new changes, it seems to me that whenever the provider is changed, the explicit build is required without any changes detection. Is it correct?

Reply:
> How do we check change for custom providers that need a rebuild?

The current entry in the properties will have a different timestamp than the persisted entry.

> it seems to me that whenever the provider is changed, the explicit build is required without any changes detection. Is it correct?

If using --optimized you will continue to use the current build - and will not see any log related to the provider change. That matches the behavior today. We certainly could add a check, just like we do for regular configuration values, and log a warning or error in this situation instead. 

If not using --optimized, it will see that an auto-build is needed.

Reply:
@shawkins right, thanks for the explanation. 

> We certainly could add a check, just like we do for regular configuration values, and log a warning or error in this situation instead.

@shawkins It might be nice to have some check for detecting changed provider. Could you please create a follow-up task? Thanks!

Reply:
@mabartos added https://github.com/keycloak/keycloak/issues/34665